rfc/router-storage.md

```diff
@@ -0,0 +1,767 @@
+      "iss": "did:web:web3.storage",
+      "aud": "did:web:web3.storage",
+      "att": {
+        "can": "/web3.storage/blob/accept",
+        "with": "did:web:web3.storage",
+        "nb": {
+          "space": "did:key:zAlice",
+          "blob": {
+            // multihash of the blob as byte array
+            "content": { "/": { "bytes": "mEi...sfKg" } },
+            "size": 2_097_152,
+          },
+          "expires": 1712903125,
+          // This task is blocked on allocation
+          "_put": { "ucan/await": [".out.ok", { "/": "bafy...put" }] }
+        }
+      }
+    }
+  }
+}
+```
+
+Note that while this appears identical to for 1.0 Invocation receipts, a careful inspection of the legacy receipt suggests something slightly different has happened: unlike the current flow, the return for `blob/allocate` is NOT synchronous. 
+
+#### Router handling of `blob/allocate`:
+
+1. The router first checks for an existing allocation and shortcuts if it exists
+2. The router runs a choice function based on the storage preferences and list of nodes available:
+
+```ts
+type ChooseNode = (prefs: StoragePreferences, nodes: StorageNodes, alreadyFailed: StorageNode[]) => StorageNode
+```
+
+**TODO: should this be a choose one or a produce complete ordering? The `alreadyFailed` argument is intended to capture previous choices that failed, to remove them from the pool, on the theory choosing 1 is probably easier and more efficient**
+
+Such a function MUST be deterministic from the input nodes and preferences. Eventually we imagine a space having a configured CID that represents a custom implementation of such a function.
+
+Let's say the chosen storage node is:
+
+```json
+{
+  "did": "did:web:node1.storage",
+  "w3upEndpoint": "https://node1.storage"
+}
+```
+
+3. The service invokes `blob/allocate` remotely on the chosen node:
+
+```json
+{ // "/": "bafy..add"
+  "cmd": "/blob/allocate",
+  "aud": "did:web:node1.storage",
+  "iss": "did:web:web3.storage",
+  "sub": "did:web:node1.storage",
+  "args": {
+    "blob": {
+      // multihash of the blob as byte array
+      "digest": { "/": { "bytes": "mEi...sfKg" } },
+      // size of the blob in bytes
+      "size": 2_097_152,
+    },
+    "storagePreferences": {
+    }
+  }
+}
+```
+
+Legacy invocation:
+
+```json
+{ // "/": "bafy..add"
+  "aud": "did:web:node1.storage",
+  "iss": "did:web:web3.storage",
+  "att": [
+    {
+      "can": "/web3.storage/blob/allocate",
+      "nb": "did:web:node1.storage",
+      "with": {
+        "blob": {
+          // multihash of the blob as byte array
+          "digest": { "/": { "bytes": "mEi...sfKg" } },
+          // size of the blob in bytes
+          "size": 2_097_152,
+        },
+        "storagePreferences": {
+        }
+      }
+    }
+  ]
+}
+```
+
+Receipt:
+
+```json
+{ 
+  // "/": "bafy..work",
+  "iss": "did:web:node1.storage",
+  "aud": "did:web:web3.storage",
+  "cmd": "/ucan/assert",
+  "sub": "did:web:web3.storage",
```

Comment:
Storage node `did:web:node1.storage` will not have an authority to sign things on `did:web:web3.storage`  sub(ject) unless explicitly authorized.  I also don't think we should authorize storage nodes to issue receipts on our behalf, instead they should issue receipts on their own sub(ject) and we can decide what to do with it.


Also in 1.0 we're going to make `iss == aud` on receipts so they form a loop and effectively provides a way to distribute state. You can think of it as I can send receipt back to the sub(ject) telling it to assert state as per receipt.

```suggestion
  "iss": "did:web:node1.storage",
  "aud": "did:web:node1.storage",
  "sub": "did:web:node1.storage",
  "cmd": "/ucan/assert",
```